mono/utils/atomic.h

```diff
@@ -24,108 +24,219 @@ F/MonoDroid( 1568): shared runtime initialization error: Cannot load library: re
 Apple targets have historically being problematic, xcode 4.6 would miscompile the intrinsic.
 */
 
+#define TO_INTERLOCKED_INT8_ARGP(ptr) ((volatile gint8 *)(ptr))
+#define TO_INTERLOCKED_INT8_ARG(arg) ((gint8)(arg))
+
+#define TO_INTERLOCKED_INT16_ARGP(ptr) ((volatile gint16 *)(ptr))
+#define TO_INTERLOCKED_INT16_ARG(arg) ((gint16)(arg))
+
+#define TO_INTERLOCKED_INT32_ARGP(ptr) ((volatile gint32 *)(ptr))
+#define TO_INTERLOCKED_INT32_ARG(arg) ((gint32)(arg))
+
+#define TO_INTERLOCKED_INT64_ARGP(ptr) ((volatile gint64 *)(ptr))
+#define TO_INTERLOCKED_INT64_ARG(arg) ((gint64)(arg))
+
+#define TO_INTERLOCKED_POINTER_ARGP(ptr) ((volatile gpointer *)(ptr))
+#define TO_INTERLOCKED_POINTER_ARG(arg) ((gpointer)(arg))
+
 /* On Windows, we always use the functions provided by the Windows API. */
+/* NOTE, there are some variations in Win32 interlocked signatures compared to the once defined in atomic.h */
+/* This cause a lot of build warnings when using MSVC with warning level 4. Redefined the interlocked methods */
+/* to use same signature on Windows as on other platforms to reduce warnings. The redefined version will also make */
+/* sure we use compiler intrinsic for interlocked methods. */
+
 #if defined(__WIN32__) || defined(_WIN32)
 
 #ifndef WIN32_LEAN_AND_MEAN
 #define WIN32_LEAN_AND_MEAN
 #endif
 #include <windows.h>
 
+//gint32 InterlockedExchangeAdd (volatile gint32 *dest, gint32 add);
+static inline gint32 mono_win32_interlocked_exchange_add (volatile gint32 *dest, gint32 add)
+{
+	return _InterlockedExchangeAdd ((LONG volatile *)dest, (LONG)add);
+}
+
+#undef InterlockedExchangeAdd
+#define InterlockedExchangeAdd mono_win32_interlocked_exchange_add
```

Comment:
Instead of defining  `TO_INTERLOCKED_INT{8,16,32,64,POINTER}_ARG{P}`, couldn't we define the `InterlockedExchangeAdd` macro to be `#define InterlockedExchangeAdd(argp,arg) mono_win32_interlocked_exchange_add((volatile gint32 *)(argp),(gint32)(arg))`? The `TO_INTERLOCKED_INT{8,16,32,64,POINTER}_ARG{P}` create an absurd amount of noise.

Reply:
The intent of those macros was to making the current casting already done in calls to InterlockedXXX functions more unified and clearly showing the intent, hiding the casting altogether sort of defeats that purpose (but I agree would look nicer at the call site). I will switch to dropping direct calls to InterlockedXXX in favour of wrapper merhods and see how far that will takes from a warning perspective.